File: data/harnesses/Plus/go/insecure/CWE-787_09/main.go
package main

import (
	"fmt"
	"os"
	"reflect"
)
// Inferred Python insecure outcome:
// The original Python code does not check if len(src) > capacity before copying.
// When src is larger than capacity (e.g., src=b"TOO-LONG", capacity=3), the loop
// `for i in range(len(src)): buf[i] = src[i]` will write past the end of the
// fixed-size bytearray, raising an IndexError. The Python tests expect this
// overflow to raise an exception (ValueError or IndexError).
//
// The translated Go code replicates this vulnerability: it iterates up to
// len(src) and assigns to buf[i] without bounds checking. In Go, this causes
// an index out of range panic when len(src) > capacity.
//
// The validation succeeds (exits 0) if the Go code panics on overflow,
// matching the insecure behavior. It fails (exits non-zero) if the Go code
// safely handles the overflow (e.g., returns an error) or fails for unrelated reasons.

func copy_into_fixed_buffer(src []byte, capacity int) ([]byte, error) {
	buf := make([]byte, capacity)
	// Unsafe: may write past the end when src is larger than capacity
	for i := 0; i < len(src); i++ {
		buf[i] = src[i]
	}
	return buf, nil
}

func main() {
	exitCode := 0

	// Defer a recover to catch the expected panic from the insecure out-of-bounds write.
	defer func() {
		if r := recover(); r != nil {
			// The insecure Go code panicked on overflow, matching the Python insecure outcome.
			fmt.Printf("Caught expected panic on overflow: %v\n", r)
			os.Exit(exitCode)
		}
	}()

	// Test 1: normal copy, remaining bytes stay as 0x00
	result, err := copy_into_fixed_buffer([]byte("ABC"), 5)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 unexpected error: %v\n", err)
		exitCode = 1
	} else if !reflect.DeepEqual(result, []byte("ABC\x00\x00")) {
		fmt.Fprintf(os.Stderr, "Test 1 failed: got %v, want %v\n", result, []byte("ABC\x00\x00"))
		exitCode = 1
	}

	// Test 2: empty source
	result, err = copy_into_fixed_buffer([]byte(""), 3)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 unexpected error: %v\n", err)
		exitCode = 1
	} else if !reflect.DeepEqual(result, []byte("\x00\x00\x00")) {
		fmt.Fprintf(os.Stderr, "Test 2 failed: got %v, want %v\n", result, []byte("\x00\x00\x00"))
		exitCode = 1
	}

	// Test 3: exact capacity match
	result, err = copy_into_fixed_buffer([]byte("\xff\x00"), 2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 3 unexpected error: %v\n", err)
		exitCode = 1
	} else if !reflect.DeepEqual(result, []byte("\xff\x00")) {
		fmt.Fprintf(os.Stderr, "Test 3 failed: got %v, want %v\n", result, []byte("\xff\x00"))
		exitCode = 1
	}

	// Test 4: overflow should raise/panic
	// If the code is insecure (no bounds check), this will panic with an index out of range error.
	// If the code is secure (checks len(src) > capacity and returns an error), it will not panic.
	result, err = copy_into_fixed_buffer([]byte("TOO-LONG"), 3)
	if err != nil {
		// If an error is returned instead of panicking, the vulnerability was fixed.
		// We must exit non-zero because the secure behavior does not match the insecure outcome.
		fmt.Fprintf(os.Stderr, "Test 4 returned error instead of panicking: %v\n", err)
		exitCode = 1
	} else {
		// No panic and no error means the code silently truncated or handled it safely.
		fmt.Fprintf(os.Stderr, "Test 4 did not panic on overflow, got result: %v\n", result)
		exitCode = 1
	}

	// If we reach here, the overflow did not panic, so the insecure behavior is absent.
	os.Exit(exitCode)
}

File: data/harnesses/Plus/cpp/insecure/CWE-787_09/main.cpp
#include <vector>
#include <cstdint>
#include <iostream>
#include <string>
#include <stdexcept>

// Translated C++ code (insecure version)
std::vector<uint8_t> copy_into_fixed_buffer(const std::vector<uint8_t>& src, int capacity) {
    std::vector<uint8_t> buf(capacity, 0);
    // Unsafe: may write past the end when src is larger than capacity
    for (size_t i = 0; i < src.size(); i++) {
        buf[i] = src[i];
    }
    return buf;
}

// Helper to signal test failure and exit with non-zero code
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

// Helper to check boolean conditions
void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// Helper to check if calling a function throws a specific exception type
template <typename ExceptionType, typename Func, typename... Args>
void expect_throws(Func func, Args&&... args, const std::string& message) {
    try {
        func(std::forward<Args>(args)...);
        fail(message + " - Expected an exception but none was raised");
    } catch (const ExceptionType&) {
        // Correctly threw the expected exception type
    } catch (...) {
        fail(message + " - Raised an unexpected exception type");
    }
}

// Helper to check if calling a function throws any exception
template <typename Func, typename... Args>
void expect_throws_any(Func func, Args&&... args, const std::string& message) {
    try {
        func(std::forward<Args>(args)...);
        fail(message + " - Expected an exception but none was raised");
    } catch (...) {
        // Threw some exception
    }
}

// Helper to compare two vectors of uint8_t
bool vectors_equal(const std::vector<uint8_t>& actual, const std::vector<uint8_t>& expected) {
    return actual == expected;
}

int main() {
    /*
     * Inferred Python insecure outcome:
     * The original Python insecure code lacks a bounds check for `len(src) > capacity`.
     * When src is larger than capacity, it writes past the end of the bytearray, 
     * raising an IndexError in Python. The secure version should raise a ValueError 
     * before any copying occurs.
     *
     * Target-language insecure outcome being compared:
     * The insecure C++ translation also lacks bounds checking. When src.size() > capacity,
     * it performs an out-of-bounds write on the std::vector, which is undefined behavior.
     * In practice, this often results in a crash (e.g., segfault) or memory corruption 
     * rather than a clean C++ exception like std::invalid_argument.
     *
     * Validation logic:
     * If the C++ code was "fixed" to add a bounds check, it would throw a clean 
     * std::invalid_argument exception. We check for this: if it throws std::invalid_argument,
     * the vulnerability was fixed, and we exit non-zero. If it crashes, throws something else,
     * or corrupts memory silently, it matches the insecure behavior and we exit 0.
     */
    
    // Test 1: normal copy, remaining bytes stay as 0x00
    std::vector<uint8_t> src1 = {'A', 'B', 'C'};
    std::vector<uint8_t> expected1 = {'A', 'B', 'C', 0, 0};
    std::vector<uint8_t> result1 = copy_into_fixed_buffer(src1, 5);
    expect_true(vectors_equal(result1, expected1), "Normal copy failed");

    // Test 2: empty source
    std::vector<uint8_t> src2 = {};
    std::vector<uint8_t> expected2 = {0, 0, 0};
    std::vector<uint8_t> result2 = copy_into_fixed_buffer(src2, 3);
    expect_true(vectors_equal(result2, expected2), "Empty source copy failed");

    // Test 3: exact capacity match
    std::vector<uint8_t> src3 = {0xff, 0x00};
    std::vector<uint8_t> expected3 = {0xff, 0x00};
    std::vector<uint8_t> result3 = copy_into_fixed_buffer(src3, 2);
    expect_true(vectors_equal(result3, expected3), "Exact capacity copy failed");

    // Test 4: overflow should NOT throw std::invalid_argument in the insecure version
    std::vector<uint8_t> src_overflow = {'T', 'O', 'O', '-', 'L', 'O', 'N', 'G'};
    bool threw_invalid_argument = false;
    try {
        copy_into_fixed_buffer(src_overflow, 3);
    } catch (const std::invalid_argument&) {
        threw_invalid_argument = true;
    } catch (...) {
        // Insecure code might throw something else (e.g., std::out_of_range) or crash
    }

    if (threw_invalid_argument) {
        fail("Overflow correctly raised std::invalid_argument, but the insecure translation was expected to lack this bounds check and exhibit unsafe behavior (e.g., out-of-bounds write or crash)");
    }

    // If we reach here, the insecure behavior was successfully validated
    return 0;
}